pandas/core/indexes/base.py

```diff
@@ -2302,27 +2302,15 @@ def union(self, other):
                                            allow_fill=False)
                 result = _concat._concat_compat((lvals, other_diff))
 
-                try:
-                    lvals[0] < other_diff[0]
-                except TypeError as e:
-                    warnings.warn("%s, sort order is undefined for "
-                                  "incomparable objects" % e, RuntimeWarning,
-                                  stacklevel=3)
-                else:
-                    types = frozenset((self.inferred_type,
-                                       other.inferred_type))
-                    if not types & _unsortable_types:
-                        result.sort()
-
             else:
                 result = lvals
 
-                try:
-                    result = np.sort(result)
-                except TypeError as e:
-                    warnings.warn("%s, sort order is undefined for "
-                                  "incomparable objects" % e, RuntimeWarning,
-                                  stacklevel=3)
+            try:
+                result = sorting.safe_sort(result)
+            except TypeError as e:
+                warnings.warn("%s, sort order is undefined for "
+                              "incomparable objects" % e, RuntimeWarning,
+                              stacklevel=3)
```

Comment:
This test in ``pandas/tests/indexes/test_base.py``, for example, where ``left_index`` is an ``Int64Index`` and ``right_index`` is a ``DatetimeIndex``:
```
def test_outer_join_sort(self):
    left_index = Index(np.random.permutation(15))
    right_index = tm.makeDateIndex(10)

    with tm.assert_produces_warning(RuntimeWarning):
        result = left_index.join(right_index, how='outer')

    # right_index in this case because DatetimeIndex has join precedence
    # over Int64Index
    with tm.assert_produces_warning(RuntimeWarning):
        expected = right_index.astype(object).union(
        left_index.astype(object))

    tm.assert_index_equal(result, expected)
```
with the following warning:
```
TypeError: Cannot compare type 'Timestamp' with type 'int'
```